### PR TITLE
[POC] core: uninstall modules using dependency based approach

### PR DESCRIPTION
**Pull request by tdecaluwe**

### CODE REVIEW

File: odoo/uninstall.py [hunk 1]
```diff
@@ -0,0 +1,1055 @@
...
+        self.env.cr.execute('''
+            SELECT included_data.id
+            FROM ir_model_data AS included_data
+            LEFT JOIN ir_model_data AS excluded_data
+            ON included_data.res_id = excluded_data.res_id
+            AND included_data.model = excluded_data.model
+            AND excluded_data.id NOT IN %s
+            WHERE included_data.id IN %s
+            AND excluded_data.id IS NULL
+        ''', (tuple(data.ids) or (None,), tuple(data.ids) or (None,)))
```
Comment (rco-odoo) [inline odoo/uninstall.py:484]:
```suggestion
        ''', (data._ids or (None,), data._ids or (None,)))
```

File: odoo/uninstall.py [hunk 2]
```diff
@@ -0,0 +1,1055 @@
...
+        relations |= Relation.search([
+            ('name', 'in', self.fields.mapped('relation_table')),
+            ('name', 'not in', Field.search([
+                ('id', 'not in', self.fields.ids),
+                ('relation_table', 'in', self.fields.mapped('relation_table')),
+                ('relation_table', '!=', False),
+            ]).mapped('relation_table')),
+        ])
+
+        if len(constraints) > 0:
```
Comment (rco-odoo) [inline odoo/uninstall.py:522]: The idiomatic way is
```suggestion
        if constraints:
```

File: addons/website_forum/models/forum.py [hunk 3]
```diff
@@ -341,7 +341,7 @@ class Post(models.Model):
     is_correct = fields.Boolean('Correct', help='Correct answer or answer accepted')
     parent_id = fields.Many2one('forum.post', string='Question', ondelete='cascade', readonly=True, index=True)
     self_reply = fields.Boolean('Reply to own question', compute='_is_self_reply', store=True)
-    child_ids = fields.One2many('forum.post', 'parent_id', string='Post Answers', domain=lambda self: [('forum_id', '=', self.forum_id.id)])
+    child_ids = fields.One2many('forum.post', 'parent_id', string='Post Answers', domain=lambda self: [('forum_id', 'in', self.forum_id.ids)])
```
[the remark this answers is not in the document]
Comment (tdecaluwe) [reply addons/website_forum/models/forum.py:344]: I added these changes to https://github.com/odoo/odoo/pull/99090

File: odoo/uninstall.py [hunk 4]
```diff
@@ -0,0 +1,1055 @@
...
+            FROM ir_model_data AS included_data
+            LEFT JOIN ir_model_data AS excluded_data
+            ON included_data.res_id = excluded_data.res_id
+            AND included_data.model = excluded_data.model
+            AND excluded_data.id NOT IN %s
+            WHERE included_data.id IN %s
+            AND excluded_data.id IS NULL
+        ''', (tuple(data.ids) or (None,), tuple(data.ids) or (None,)))
+        data = data.browse(rec[0] for rec in self.env.cr.fetchall())
+        data = data.sorted(key='model')
```
[the remark this answers is not in the document]
Comment (tdecaluwe) [reply odoo/uninstall.py:486]: So the extra `included_data.id` in the ORDER BY clause is only there to remove the nondeterminism? What makes more sense: `included_data.id` or `included_data.res_id`?

**Pull request closed without merging**